File: Easy_Level_using_Cpp/121. Best Time to Buy and Sell Stock.cpp
// Problem - 121. Best Time to Buy and Sell Stock
// Link - https://leetcode.com/problems/best-time-to-buy-and-sell-stock/description/
// Runtime 0ms Beats 100.00% Analyze Complexity Memory 96.00 MB Beats 98.69%
// Solution : 
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int minPrice = prices[0];
        int maxProfit = 0;

        for(int i = 1; i < prices.size(); i++){
            if(prices[i] < minPrice){
                minPrice = prices[i];
            } else if(prices[i] - minPrice > maxProfit){
                maxProfit = prices[i] - minPrice;
            }
        }
        return maxProfit;
    }
};

File: Easy_Level_using_Python/121. Best Time to Buy and Sell Stock.py
# Problem - 121. Best Time to Buy and Sell Stock
# Link - https://leetcode.com/problems/best-time-to-buy-and-sell-stock/description/
# Runtime 42 ms Beats 80.26% Analyze Complexity Memory 25.98 MB Beats 78.93%
# Time Complexity: O(n)
# Space Complexity: O(1)
# Solution : 
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        min_price = prices[0]
        best_price = 0
        for i in range(1, len(prices)):
            new_price = prices[i] - min_price
            if best_price < new_price:
                best_price = new_price
            if min_price > prices[i]:
                min_price = prices[i]
        return best_price
